File: ASPNETCore/HowTo/WebApiConsoleSample/src/Modules/DataEngine/DataEngine.Controller.cs
using System;
using System.Net.Http;
using System.Threading;
using Newtonsoft.Json.Linq;

namespace WebApiConsoleSample
{

    public class DataEngineController : BaseController<DataEngineService>
    {
        public DataEngineController(DataEngineService service) : base(service)
        {

        }
        public override void Run()
        {
            // GetAllFields();
            AnalyzeData();
        }

        private void GetAllFields()
        {

            Console.WriteLine("\n========= GetAllFields started ===========\n");

            GetAllFieldDto dto = new GetAllFieldDto();
            dto.DataSource = "complex10";

            service.GetAllFields(dto).Wait();

            Console.WriteLine("\n========= GetAllFields ended =============\n");
        }

        private void AnalyzeData()
        {
            Console.WriteLine("\n========= AnalyzeData started ===========\n");
            AnalyzeDataDto dto = new AnalyzeDataDto();
            dto.DataSource = "complex10";
            JObject obj = new JObject();
            obj.Add("view", "{\"showZeros\":false,\"showColumnTotals\":0,\"showRowTotals\":0,\"defaultFilterType\":1,\"totalsBeforeData\":false,\"sortableGroups\":true,\"fields\":[{\"binding\":\"Active\",\"header\":\"Active\",\"dataType\":3,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Active\"},{\"binding\":\"Country\",\"header\":\"Country\",\"dataType\":1,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Country\"},{\"binding\":\"Date\",\"header\":\"Date\",\"dataType\":4,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"d\",\"isContentHtml\":false,\"key\":\"Date\"},{\"binding\":\"Discount\",\"header\":\"Discount\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Discount\"},{\"binding\":\"Downloads\",\"header\":\"Downloads\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Downloads\"},{\"binding\":\"ID\",\"header\":\"ID\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"ID\"},{\"binding\":\"Product\",\"header\":\"Product\",\"dataType\":1,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Product\"},{\"binding\":\"Sales\",\"header\":\"Sales\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Sales\"}],\"rowFields\":{\"items\":[\"Product\",\"Country\"]},\"columnFields\":{\"items\":[]},\"filterFields\":{\"items\":[]},\"valueFields\":{\"items\":[\"Sales\",\"Downloads\"]}}");
            dto.ViewDefinition = obj.ToString();
            var response = service.AnalyzeData(dto).Result;
            Console.WriteLine("AnalyzeData getting response: {0}", response);
            Console.WriteLine("\n========= AnalyzeData ended =============\n");

            bool IsCompleted = false;
            while (!IsCompleted)
            {
                Thread.Sleep(1000);
                string analysisStatus = AnalyzeStatus(response);
                try{
                     JObject analysisStatusObj = JObject.Parse(analysisStatus);
                     if(analysisStatusObj != null){
                         if(analysisStatusObj.ContainsKey("executingStatus") && analysisStatusObj.GetValue("executingStatus").ToString().Equals("Completed")){
                             IsCompleted = true;
                         }
                     }
                }catch(Exception e){
                    Console.WriteLine(e.StackTrace);
                    IsCompleted = true;
                }
            }

            AnalyzeResult(response);
            return;
        }


        private void AnalyzeResult(string analyDataResponse)
        {
            try
            {
                Console.WriteLine("\n========= AnalyzeResult started ===========\n");
                JObject obj = JObject.Parse(analyDataResponse);

                AnalyzeResultDto dto = new AnalyzeResultDto();
                dto.DataSource = "complex10";
                dto.Token = obj.GetValue("token").ToString();
                var result = service.AnalyzeResult(dto).Result;
                Console.WriteLine("Analyzed with result {0}", result);
                Console.WriteLine("\n========= AnalyzeResult ended =============\n");
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private string AnalyzeStatus(string analyDataResponse)
        {
            try
            {
                Console.WriteLine("\n========= AnalyzeStatus started ===========\n");
                JObject obj = JObject.Parse(analyDataResponse);

                AnalyzeStatusDto dto = new AnalyzeStatusDto();
                dto.DataSource = "complex10";
                dto.Token = obj.GetValue("token").ToString();
                var result = service.AnalyzeStatus(dto).Result;
                Console.WriteLine("AnalyzeData getting response {0}", result);
                Console.WriteLine("\n========= AnalyzeStatus ended =============\n");
                return result;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
                return null;
            }
        }
    }
}

File: ASPNETCore/HowTo/WebApiTypescriptSample/src/Modules/DataEngine/DataEngine.Controller.ts
import { BaseController } from '../../Base/Base.Controller';
import * as FSExtra from 'fs-extra';
import * as Path from 'path';
import { DataEngineService } from './DataEngine.Service';
import { GetAllFieldsDto } from './Dtos/GetAllFields.dto';
import { AnalyzeDataDto } from './Dtos/AnalyzeData.dto';
import { AnalyzeResultDto } from './Dtos/AnalyzeResult.dto';
import { AnalyzeStatusDto } from './Dtos/AnalyzeStatus.dto';
export class DataEngineController extends BaseController<DataEngineService>{

    constructor() {
        super(new DataEngineService());
    }

    public async Run(): Promise<any> {
        // await this.GetAllFields();
        this.AnalyzeData();
    }


    private async GetAllFields() {
        console.log("\n========= GetAllFields started ===========\n");
        let dto: GetAllFieldsDto = new GetAllFieldsDto();
        dto.DataSource = "complex10";
        let res = await this.service.GetAllFields(dto);
        console.log(res);
        console.log("\n========= GetAllFields ended =============\n");
    }

    private async AnalyzeData() {
        console.log("\n========= AnalyzeData started ===========\n");
        let dto: AnalyzeDataDto = new AnalyzeDataDto();
        dto.DataSource = "complex10";
        dto.ViewDefinition = JSON.stringify({
            view: "{\"showZeros\":false,\"showColumnTotals\":0,\"showRowTotals\":0,\"defaultFilterType\":1,\"totalsBeforeData\":false,\"sortableGroups\":true,\"fields\":[{\"binding\":\"Active\",\"header\":\"Active\",\"dataType\":3,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Active\"},{\"binding\":\"Country\",\"header\":\"Country\",\"dataType\":1,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Country\"},{\"binding\":\"Date\",\"header\":\"Date\",\"dataType\":4,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"d\",\"isContentHtml\":false,\"key\":\"Date\"},{\"binding\":\"Discount\",\"header\":\"Discount\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Discount\"},{\"binding\":\"Downloads\",\"header\":\"Downloads\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Downloads\"},{\"binding\":\"ID\",\"header\":\"ID\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"ID\"},{\"binding\":\"Product\",\"header\":\"Product\",\"dataType\":1,\"aggregate\":2,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Product\"},{\"binding\":\"Sales\",\"header\":\"Sales\",\"dataType\":2,\"aggregate\":1,\"showAs\":0,\"descending\":false,\"format\":\"n0\",\"isContentHtml\":false,\"key\":\"Sales\"}],\"rowFields\":{\"items\":[\"Product\",\"Country\"]},\"columnFields\":{\"items\":[]},\"filterFields\":{\"items\":[]},\"valueFields\":{\"items\":[\"Sales\",\"Downloads\"]}}"
        });
        let res = await this.service.AnalyzeData(dto);
        console.log("AnalyzeData getting response: ", res.data);
        console.log("\n========= AnalyzeData ended =============\n");
        
        let IsCompleted: boolean = false;
        while (!IsCompleted)
        {
            await this.delay(1000);
            let analysisStatus = await this.AnalyzeStatus(res.data);
            let responseObj = analysisStatus.data;
            try{
                
                 if(responseObj){
                     if(responseObj.hasOwnProperty("executingStatus") && responseObj['executingStatus'] == "Completed"){
                         IsCompleted = true;
                     }
                 }
            }catch(e){    
                IsCompleted = true;
            }
        }
        this.AnalyzeResult(res.data);
    }
    async delay(ms: number) {
        await new Promise(resolve => setTimeout(() => resolve(), ms)).then(() => { });
    }

    private async AnalyzeResult(res: any) {
        try {
            console.log("\n========= AnalyzeResult started ===========\n");
            let dto: AnalyzeResultDto = new AnalyzeResultDto();
            dto.DataSource = "complex10";
            dto.Token = res["token"];
            var result = await this.service.AnalyzeResult(dto);
            console.log("Analyzed with result ", result.data);
            console.log("\n========= AnalyzeResult ended =============\n");
        }
        catch (error) {
            console.log(error);
            return Promise.resolve(null);
        }
    }

    private async AnalyzeStatus(res: any) {
        try {

            console.log("\n========= AnalyzeStatus started ===========\n");
            let dto: AnalyzeStatusDto = new AnalyzeStatusDto();
            dto.DataSource = "complex10";
            dto.Token = res["token"];
            var result = await this.service.AnalyzeStatus(dto);
            console.log("AnalyzeData getting response ", result.data);
            console.log("\n========= AnalyzeStatus ended =============\n");
            return result;
        }
        catch (e) {
            return null;
        }
    }
}